client/js/chart/index.js

```diff
@@ -0,0 +1,131 @@
+import {SVG_NS} from '../util.js';
+
+/**
+ * Creates an SVG chart inside of `el` with the given data.
+ *
+ * @param {HTMLElement} el The container element for this chart.
+ * @param {Map<string, number[]>} rankingHistory The ranking history for each
+ * track.
+ * @param {Date[]} rankingDates The date of each history entry.
+ */
+export function createChart(el, rankingHistory, rankingDates) {
+  const svg = document.createElementNS(SVG_NS, 'svg');
+  svg.setAttribute('class', 'chart');
+  svg.append(createDefs());
+
+  const seriesContainer = document.createElementNS(SVG_NS, 'g');
+  seriesContainer.setAttribute('class', 'series-container');
+  svg.append(seriesContainer);
+
+  let index = 0;
+  const colors = ['blue', 'red', 'green', 'purple', 'orange'];
+
+  // get only the first 7 tracks for now
+  for (const history of [...rankingHistory.values()].slice(0, 7)) {
+    const color = colors[index % colors.length];
+    const series = createSeries(color, history);
+    seriesContainer.append(series);
+    index++;
+  }
+
+  el.append(svg);
+}
+
+const RUN_SCALE_X = 30;
+const RUN_SCALE_Y = 30;
+
+/**
+ * Creates a new series (set of lines on the chart for a specific song).
+ *
+ * @param {string} color The color of this series.
+ * @param {number[]} history The historical positions of this track on the
+ * leaderboard.
+ * @returns {SVGGElement} A group containing the series.
+ */
+function createSeries(color, history) {
+  const series = document.createElementNS(SVG_NS, 'g');
+  series.setAttribute('class', 'series');
+  series.style.setProperty('--run-color', color);
+  let start = 0;
+  let end = 0;
+  while (end < history.length) {
+    while (start < history.length && history[start] === null) {
+      start++;
+    }
+    end = start + 1;
+    while (end < history.length && history[end] !== null) {
+      end++;
+    }
+    series.append(...createRun(history, start, end));
+    start = end;
+  }
+
+  return series;
+}
+
+/**
+ * Creates a new run (continuous group of points within a series).
+ *
+ * @param {number[]} history The historical positions of this track on the
+ * leaderboard.
+ * @param {number} start The index of the first entry in this run.
+ * @param {number} end The index of the last entry in this run.
+ * @returns {SVGElement[]} The elements that compose this run.
+ */
+function createRun(history, start, end) {
+  const points = history
```

Comment:
It'll be more readable if you separate this out into 2 variables: points and pointsStr. The points are just the x/y pairs for each of these points, whereas pointsStr is the concatedated pair in "x,y x,y..." format. This will help when you create the end caps since the code becomes:
```
startCap.setAttribute('cx', points[0].x + 'px');
startCap.setAttribute('cy', points[0].y + 'px');
```